File: pilot_center-go/room/room.go
package room

import (
	"encoding/json"
	"pilot_center-go/logger"
	"sync"
)

// Room 表示WebSocket通信中的房间
type Room struct {
	RoomID   string
	logger   *logger.Logger
	users    map[string]*User
	sessions map[string]interface{}
	usersMu  sync.RWMutex
	sessionsMu sync.RWMutex
}

// NewRoom 创建一个新的房间
func NewRoom(roomID string) *Room {
	return &Room{
		RoomID:   roomID,
		logger:   logger.GetLogger(),
		users:    make(map[string]*User),
		sessions: make(map[string]interface{}),
	}
}

// AddUser 添加用户到房间
func (r *Room) AddUser(user *User) bool {
	r.usersMu.Lock()
	defer r.usersMu.Unlock()

	if _, exists := r.users[user.UserID]; exists {
		return false
	}

	r.users[user.UserID] = user
	return true
}

// RemoveUser 从房间移除用户
func (r *Room) RemoveUser(userID string) bool {
	r.usersMu.Lock()
	defer r.usersMu.Unlock()

	if _, exists := r.users[userID]; exists {
		delete(r.users, userID)
		return true
	}

	return false
}

// GetUser 获取房间中的用户
func (r *Room) GetUser(userID string) *User {
	r.usersMu.RLock()
	defer r.usersMu.RUnlock()

	return r.users[userID]
}

// ListUsers 列出房间中的所有用户
func (r *Room) ListUsers() []*User {
	r.usersMu.RLock()
	defer r.usersMu.RUnlock()

	result := make([]*User, 0, len(r.users))
	for _, user := range r.users {
		result = append(result, user)
	}

	return result
}

// UserCount 获取房间中的用户数量
func (r *Room) UserCount() int {
	r.usersMu.RLock()
	defer r.usersMu.RUnlock()

	return len(r.users)
}

// AddSession 添加会话到房间
func (r *Room) AddSession(peer string, session interface{}) {
	r.sessionsMu.Lock()
	defer r.sessionsMu.Unlock()

	r.sessions[peer] = session
}

// RemoveSession 从房间移除会话
func (r *Room) RemoveSession(peer string) {
	r.sessionsMu.Lock()
	defer r.sessionsMu.Unlock()

	delete(r.sessions, peer)
}

// HandleJoin 处理用户加入房间
func (r *Room) HandleJoin(userID, userName string, audience bool, session interface{}) interface{} {
	r.logger.Infof("用户 %s(%s) 加入房间 %s", userID, userName, r.RoomID)

	// 检查用户是否已存在
	existing := r.GetUser(userID)
	if existing == nil {
		// 创建新用户
		user := NewUser(userID, userName)
		r.AddUser(user)
		existing = user
	} else {
		r.logger.Infof("用户 %s 已存在于房间 %s", userID, r.RoomID)
	}

	// 关联会话
	if session != nil {
		existing.AddSession(session)
		// 注册会话到房间
		if getPeer, ok := session.(interface {
			GetPeer() string
		}); ok {
			peer := getPeer.GetPeer()
			r.AddSession(peer, session)
		}
	}

	// 通知其他用户新用户加入
	if !audience {
		r.NotifyNewUser(userID, userName, userID)
	}

	// 构建响应数据
	responseData := map[string]interface{}{
		"code":    0,
		"message": "join success",
		"roomId":  r.RoomID,
		"users":   []map[string]interface{}{},
	}

	// 添加其他用户信息
	for _, user := range r.ListUsers() {
		// 跳过当前用户
		if user.UserID == userID {
			continue
		}

		userInfo := map[string]interface{}{
			"userId":   user.UserID,
			"userName": user.Name,
			"pushers":  []map[string]interface{}{},
		}

		// 添加推流信息
		pushers := user.GetPusherInfo()
		for _, pusher := range pushers {
			userInfo["pushers"] = append(userInfo["pushers"].([]map[string]interface{}), pusher.ToDict())
		}

		responseData["users"] = append(responseData["users"].([]map[string]interface{}), userInfo)
	}

	return responseData
}

// BroadcastExceptUser 向房间中的所有用户广播通知，除了指定的用户
func (r *Room) BroadcastExceptUser(method string, payload map[string]interface{}, exceptUserID string) {
	// 获取要排除的会话
	var excludedSession interface{}
	if exceptUserID != "" {
		user := r.GetUser(exceptUserID)
		if user != nil {
			excludedSession = user.GetSession()
		}
	}

	// 广播通知
	r.sessionsMu.RLock()
	defer r.sessionsMu.RUnlock()

	for _, session := range r.sessions {
		// 跳过排除的会话
		if session == excludedSession {
			continue
		}

		// 发送通知
		if sendNotification, ok := session.(interface {
			SendNotification(method string, data map[string]interface{}) error
		}); ok {
			if err := sendNotification.SendNotification(method, payload); err != nil {
				r.logger.Errorf("发送通知失败: %v", err)
			}
		}
	}
}

// NotifyNewUser 通知房间中的其他用户有新用户加入
func (r *Room) NotifyNewUser(userID, userName, exceptUserID string) {
	payload := map[string]interface{}{
		"roomId":   r.RoomID,
		"userId":   userID,
		"userName": userName,
	}

	r.logger.Infof("通知房间 %s 成员新用户 %s(%s) 加入", r.RoomID, userID, userName)
	r.BroadcastExceptUser("newUser", payload, exceptUserID)
}

// HandleUserDisconnect 处理用户断开连接
func (r *Room) HandleUserDisconnectNotification(data map[string]interface{}, session interface{}) {
	userID, _ := data["userId"].(string)
	r.logger.Infof("处理用户 %s 在房间 %s 的断开连接通知", userID, r.RoomID)

	// 获取用户
	user := r.GetUser(userID)
	if user == nil {
		r.logger.Infof("用户 %s 在房间 %s 中不存在", userID, r.RoomID)
		return
	}

	// 通知其他用户
	payload := map[string]interface{}{
		"roomId": r.RoomID,
		"userId": userID,
	}
	r.BroadcastExceptUser("userDisconnect", payload, userID)

	// 移除会话关联
	user.RemoveSession(session)
}

// HandleUserLeave 处理用户离开房间
func (r *Room) HandleUserLeaveNotification(data map[string]interface{}, session interface{}) {
	userID, _ := data["userId"].(string)
	r.logger.Infof("处理用户 %s 在房间 %s 的离开通知", userID, r.RoomID)

	// 获取用户
	user := r.GetUser(userID)
	if user == nil {
		r.logger.Infof("用户 %s 在房间 %s 中不存在", userID, r.RoomID)
		return
	}

	// 通知其他用户
	payload := map[string]interface{}{
		"roomId": r.RoomID,
		"userId": userID,
	}
	r.BroadcastExceptUser("userLeave", payload, userID)

	// 移除会话关联
	user.RemoveSession(session)

	// 从房间移除用户
	r.RemoveUser(userID)
}

// HandleTextMessage 处理文本消息
func (r *Room) HandleTextMessageNotification(data map[string]interface{}, session interface{}) {
	userID, _ := data["userId"].(string)
	userName, _ := data["userName"].(string)
	message, _ := data["message"].(string)

	r.logger.Infof("处理用户 %s(%s) 在房间 %s 的文本消息: %s", userID, userName, r.RoomID, message)

	// 广播文本消息
	r.BroadcastExceptUser("textMessage", map[string]interface{}{
		"roomId":   r.RoomID,
		"userId":   userID,
		"userName": userName,
		"message":  message,
	}, userID)
}

// HandlePushNotification 处理推流通知
func (r *Room) HandlePushNotification(data map[string]interface{}, session interface{}) {
	userID, _ := data["userId"].(string)
	userName, _ := data["userName"].(string)
	publishers, _ := data["publishers"].([]interface{})

	r.logger.Infof("在房间 %s 中收到用户 %s(%s) 的推流通知，发布者: %v", r.RoomID, userID, userName, publishers)

	// 获取用户
	user := r.GetUser(userID)
	if user == nil {
		r.logger.Infof("在房间 %s 中不存在用户 %s", r.RoomID, userID)
		// 创建新用户
		user = NewUser(userID, userName)
		r.AddUser(user)
	} else {
		// 更新用户名
		if userName != "" {
			user.Name = userName
		}
	}

	// 处理发布者信息
	var pushInfoList []*PushInfo
	for _, pub := range publishers {
		if pubDict, ok := pub.(map[string]interface{}); ok {
			pusherID, _ := pubDict["pusherId"].(string)
			if pusherID == "" {
				continue
			}

			// 创建推流信息
			pushInfo := &PushInfo{PusherID: pusherID}

			// 解析RTP参数
			if rtpParamDict, ok := pubDict["rtpParam"].(map[string]interface{}); ok {
				rtpParam := &RtpParam{}
				rtpParam.FromDict(rtpParamDict)
				pushInfo.SetRtpParam(rtpParam)
			}

			// 添加到推流列表
			pushInfoList = append(pushInfoList, pushInfo)
			// 更新用户的推流信息
			user.SetPusherInfo(pushInfo)
		}
	}

	// 通知其他用户新的推流
	r.NotifyNewPushers(userID, user.Name, pushInfoList)
}

// NotifyNewPushers 通知房间中的其他用户有新的推流
func (r *Room) NotifyNewPushers(userID, userName string, pushInfoList []*PushInfo) {
	// 构建推流信息列表
	pushersList := []map[string]interface{}{}
	for _, pushInfo := range pushInfoList {
		pushersList = append(pushersList, pushInfo.ToDict())
	}

	// 广播通知
	payload := map[string]interface{}{
		"roomId":   r.RoomID,
		"userId":   userID,
		"userName": userName,
		"pushers":  pushersList,
	}

	r.BroadcastExceptUser("newPusher", payload, userID)
}

// HandlePullRemoteStream 处理拉取远程流通知
func (r *Room) HandlePullRemoteStreamNotification(data map[string]interface{}, session interface{}) {
	pusherUserID, _ := data["pusher_user_id"].(string)

	// 发送拉取请求给推流用户
	pusherUser := r.GetUser(pusherUserID)
	if pusherUser == nil {
		r.logger.Infof("在房间 %s 中不存在推流用户 %s", r.RoomID, pusherUserID)
		return
	}

	pusherSession := pusherUser.GetSession()
	if pusherSession == nil {
		r.logger.Infof("推流用户 %s 在房间 %s 中没有会话", pusherUserID, r.RoomID)
		return
	}

	// 记录数据
	dataBytes, _ := json.Marshal(data)
	r.logger.Infof("向房间 %s 中的推流用户 %s 发送拉取远程流通知: %s", r.RoomID, pusherUserID, string(dataBytes))

	// 发送通知
	if sendNotification, ok := pusherSession.(interface {
		SendNotification(method string, data map[string]interface{}) error
	}); ok {
		if err := sendNotification.SendNotification("pullRemoteStream", data); err != nil {
			r.logger.Errorf("发送拉取远程流通知失败: %v", err)
		}
	}
}

File: pilot_center/room/room.py
"""Room container.

Room holds multiple User instances and provides simple operations to add,
remove and list users. This is deliberately small and synchronous; the
pilot_center server will call these methods from asyncio tasks so avoid
blocking operations here.
"""

from __future__ import annotations

import logging
import json
from typing import Dict, List, Optional
from .user import User
from .push_info import PushInfo
from .rtp_param import RtpParam

class Room:
	def __init__(self, room_id: str) -> None:
		"""Initialize a Room identified by room_id.

		Note: `name` field was removed — callers should use `room_id` as the
		canonical identifier if a human-friendly name is not available.
		"""
		self.room_id = room_id
		# per-room logger for easier filtering in logs
		self.log = logging.getLogger(f"room.{room_id}")
		# Log creation and constructor input parameters for observability
		self.log.info("Room created: room_id=%s, init_args=%s", room_id, {"room_id": room_id})
		# users keyed by user_id
		self._users: Dict[str, User] = {}

		# sessions keyed by session id (e.g. ip:port) for room-level broadcasts
		# value is the session object (WsProtooSession)
		self._sessions: Dict[str, object] = {}

	def add_user(self, user: User) -> bool:
		"""Add user to room. Return True if added, False if already present."""
		if user.user_id in self._users:
			return False
		self._users[user.user_id] = user
		return True

	def remove_user(self, user_id: str) -> bool:
		"""Remove user by id. Return True if removed, False if not found."""
		return self._users.pop(user_id, None) is not None

	def get_user(self, user_id: str) -> Optional[User]:
		return self._users.get(user_id)

	def list_users(self) -> List[User]:
		return list(self._users.values())

	def user_count(self) -> int:
		return len(self._users)

	def handle_join(self, user_id: str, user_name: str = "", audience: bool = False, session: object | None = None) -> dict:
		"""Handle a user joining this room.

		Creates a User if missing, associates the (single) session with the
		User and registers the session in the room's session map (keyed by
		session.peer if available). Returns True if the user was added, False
		if the user was already present.
		"""
		self.log.info("User %s(%s) joining room %s", user_id, user_name, self.room_id)
		existing = self.get_user(user_id)
		if existing is None:
			user = User(user_id=user_id, name=user_name)
			self.add_user(user)
		else:
			self.log.info("User %s already exists in room %s", user_id, self.room_id)
			user = existing

		if session is not None:
			# associate session with the user (User manages single session)
			try:
				user.add_session(session)
			except Exception:
				pass
			# register session under this room for broadcast lookups
			sid = getattr(session, "peer", None)
			if sid:
				try:
					self.add_session(sid, session)
				except Exception:
					pass
		# log the self.sessions size
		self.log.info("Room %s has %d registered sessions after user %s join", self.room_id, len(self._sessions), user_id)
		# call self.notify_new_user
		# notify other room members that a new user joined
		try:
			if not audience:
				self.notify_new_user(user_id, user_name, except_userId=user_id)
		except Exception:
			pass
		data = {
			"code": 0,
			"message": "join success",
			"roomId": self.room_id,
			"users": [],
		}
		for u in self.list_users():
			# skip the joining user
			if u.user_id == user_id:
				continue
			user_info = {
				"userId": u.user_id,
				"userName": u.name,
				"pushers": [],
			}
			pushers = u.get_pusher_info()
			for p in pushers.values():
				user_info["pushers"].append(p.to_dict())
			#skip pushers for now
			data["users"].append(user_info)

		return data

	def broadcast(self, method: str, payload: Optional[Dict] = None) -> None:
		"""Stub: broadcast a message to all users in the room.

		In this pilot_center prototype we don't have a real messaging channel
		(WebSocket sessions live in other modules). This method exists as a
		convenient hook — callers may iterate users and deliver messages.
		"""
		# Broadcast to all sessions registered in this room (best-effort).
		for sess in list(self._sessions.values()):
			try:
				# session should implement async send_notification; schedule fire-and-forget
				import asyncio as _asyncio
				_asyncio.create_task(sess.send_notification(method, payload))
			except Exception:
				# ignore individual session failures
				pass

	def broadcast_except_user(self, message: str, userId: str, method: str = "room.broadcast", payload: Optional[Dict] = None) -> None:
		"""Broadcast to all sessions in this room except the session belonging
		to `userId`.

		Backward compatible: when `payload` is None, this sends
		`{"message": message}` with method `method` (default "room.broadcast").
		If `payload` is provided, that dict is sent as the data for `method`.

		This uses the per-room session map (`_sessions`) and excludes the
		session returned by `User.get_session()` for `userId` (single-session
		model). The operation is best-effort (individual send failures are
		ignored).
		"""
		excluded_session = None
		user = self.get_user(userId)
		if user is not None:
			# determine the single session object to exclude (single-session model)
			try:
				excluded_session = user.get_session()
			except Exception:
				excluded_session = None

		data = payload if payload is not None else {"message": message}

		for sess in list(self._sessions.values()):
			# skip the excluded session (those belonging to the provided userId)
			if sess is excluded_session:
				continue
			try:
				import asyncio as _asyncio
				self.log.info("Broadcasting to session %s method=%s data=%s", getattr(sess, "peer", "unknown"), method, data)
				_asyncio.create_task(sess.send_notification(method, data))
			except Exception:
				# best-effort: ignore send failures
				self.log.info("Failed to send to session %s", getattr(sess, "peer", "unknown"))
				pass

	def add_session(self, session_id: str, session: object) -> None:
		"""Register a session under this room (for room-level broadcasts)."""
		if not session_id:
			return
		self._sessions[session_id] = session

	def remove_session(self, session_id: str) -> None:
		"""Remove a session previously registered for this room."""
		try:
			self._sessions.pop(session_id, None)
		except Exception:
			pass

	def notify_new_user(self, userId: str, name: str, except_userId: Optional[str] = None) -> None:
		"""Notify room members that a new user joined.

		Sends a notification with method `newUser` and data:
		{ "roomId": self.room_id, "userId": userId, "userName": name }

		If `except_userId` is provided, that user's session (if any) will be
		excluded from the broadcast.
		"""

		payload = {"roomId": self.room_id, "userId": userId, "userName": name}
		self.log.info("Notifying room %s members of new user %s(%s), except %s", self.room_id, userId, name, except_userId)
		# reuse broadcast_except_user to avoid duplicating the broadcast loop
		# pass the payload and method so it sends the correct notification shape
		# call broadcast_except_user(self, message: str, userId: str, method: str = "room.broadcast", payload: Optional[Dict] = None) -> None:
		self.broadcast_except_user("", except_userId or "", method="newUser", payload=payload)

	def handle_userDisconnect_notification(self, data: Dict[str, object], session: object) -> None:
		"""Handle a userDisconnect notification sent to this room.

		Removes the session from the room and disassociates it from the user.
		"""
		user_id = data.get("userId", "")
		self.log.info("Handling userDisconnect notification for user %s in room %s", user_id, self.room_id)
		user = self.get_user(user_id)
		if user is None:
			self.log.info("userDisconnect for unknown user %s in room %s", user_id, self.room_id)
			return

		self.log.info("userDisconnect completed for user %s in room %s", user_id, self.room_id)
		payload = {"roomId": self.room_id, "userId": user_id}
		self.broadcast_except_user("", user_id, method="userDisconnect", payload=payload)
		# remove session from user
		try:
			user.remove_session(session)
		except Exception:
			pass
		# donn't remove session from room for there may be other users using the same session
	def handle_userLeave_notification(self, data: Dict[str, object], session: object) -> None:
		"""Handle a userLeave notification sent to this room.

		Removes the user from the room and disassociates the session.
		"""
		user_id = data.get("userId", "")
		self.log.info("Handling userLeave notification for user %s in room %s", user_id, self.room_id)
		user = self.get_user(user_id)
		if user is None:
			self.log.info("userLeave for unknown user %s in room %s", user_id, self.room_id)
			return

		self.log.info("userLeave completed for user %s in room %s", user_id, self.room_id)
		payload = {"roomId": self.room_id, "userId": user_id}
		self.broadcast_except_user("", user_id, method="userLeave", payload=payload)
		# remove session from user
		try:
			user.remove_session(session)
		except Exception:
			pass
		# remove user from room
		try:
			self.remove_user(user_id)
		except Exception:
			pass

	def handle_textMessage_notification(self, data: Dict[str, object], session: object) -> None:
		"""Handle a textMessage notification sent to this room.

		Broadcasts the text message to all room members.
		"""
		user_id = data.get("userId", "")
		user_name = data.get("userName", "")
		message = data.get("message", "")
		self.log.info("Handling textMessage notification from user %s(%s) in room %s: %s", user_id, user_name, self.room_id, message)
		# broadcast the message to all room members
		self.broadcast_except_user("", user_id, method="textMessage", payload={
			"roomId": self.room_id,
			"userId": user_id,
			"userName": user_name,
			"message": message,
		})

	def handle_push_notification(self, data: Dict[str, object], session: object) -> None:
		"""Handle a push notification sent to this room.

		For now this is a stub that logs the data.
		"""
		pusher_user_id = data.get("userId", "")
		push_user_name = data.get("userName", "")
		publishers = data.get("publishers", [])
		self.log.info("Received push notification in room %s from user %s(%s) with publishers: %s",
			self.room_id, pusher_user_id, push_user_name, publishers)
		# Further processing can be implemented here as needed.

		user = self.get_user(pusher_user_id)
		if user is None:
			self.log.info("Push notification from unknown user %s in room %s", pusher_user_id, self.room_id)
			new_user = User(user_id=pusher_user_id, name=push_user_name)
			self.add_user(new_user)
			user = new_user
		push_user_name = user.get_name()
		# For now, we just log the pushers info.
		push_info_list = []
		for pub in publishers:
			rtpParam = pub.get("rtpParam", {})
			pusherId = pub.get("pusherId", "")
			push_info = PushInfo(pusherId=pusherId)
			push_info.set_rtp_param(RtpParam.from_dict(rtpParam))

			push_info_list.append(push_info)
			user.set_pusher_info(push_info)

		self.notify_new_pushers(pusher_user_id, push_user_name, push_info_list)

	def notify_new_pushers(self, userId: str, userName: str, push_info_list: List[PushInfo]) -> None:
		"""Notify room members that a user has new pushers.

		Sends a notification with method `newPusher` and data:
		{ "roomId": self.room_id, "userId": userId, "pushers": pushers }
		"""
		pushers_list = []
		for p in push_info_list:
			pushers_list.append({
				"pusherId": p.pusherId,
				"rtpParam": p.rtpParam.to_dict() if p.rtpParam else {},
			})
		
		payload = {"roomId": self.room_id, "userId": userId, "userName": userName, "pushers": pushers_list}
		self.broadcast_except_user("", userId, method="newPusher", payload=payload)

	def handle_pull_remote_stream_notification(self, data: Dict[str, object], session: object) -> None:
		"""Handle a pull remote stream notification sent to this room.

		For now this is a stub that logs the data.
		"""
		pusher_user_id = data.get("pusher_user_id", "")

		# send pull request to the pusher user in sfu
		pusher_user = self.get_user(pusher_user_id)
		if pusher_user is None:
			self.log.info("Pull remote stream notification for unknown pusher user %s in room %s", pusher_user_id, self.room_id)
			return
		session = pusher_user.get_session()
		if session is None:
			self.log.info("Pull remote stream notification for pusher user %s with no session in room %s", pusher_user_id, self.room_id)
			return
		import asyncio as _asyncio
		data_str = json.dumps(data)
		self.log.info("Sending pullRemoteStream notification to pusher user %s in room %s, data:%s", pusher_user_id, self.room_id, data_str)
		_asyncio.create_task(session.send_notification("pullRemoteStream", data))

	def handle_asr_result_notification(self, data: Dict[str, object], session: object) -> None:
		"""Handle an asr_result notification sent to this room.

		Broadcasts the asr result to all room members.
		"""

		"""
		{
		    "type": "asr_result",
		    "result": "I am a test ASR result",
		    "start_ms": "17000000",
		    "end_ms": "17111111",
		    "index": 1,
		    "ts": "1712344444",
		    "roomId": "abcde",
		    "userId": "eeeee",
		    "userName": "eeeee_username"
		}
		handle the asr_result notification sent to this room.
		"""
		user_id = data.get("userId", "")
		user_name = data.get("userName", "")
		asr_result = data.get("result", "")

		asr_str = user_name + ": " + asr_result

		self.log.info("Handling asr_result notification from user %s(%s) in room %s: %s", user_id, user_name, self.room_id, asr_result)
		# broadcast the asr result to all room members
		self.broadcast(method="textMessage", payload={
			"roomId": self.room_id,
			"userId": 'ai_asr_bot',
			"userName": 'AI_ASR_Bot',
			"message": asr_str,
		})
